src/common/pf_alarm.c

```diff
@@ -2267,74 +2236,296 @@ static int pf_alarm_send_internal (
- * @param p_ar             InOut: The AR instance.
- * @return 0 is always returned
+ * @internal
+ * Create a mutex for a queue.
+ *
+ * Note that on first usage, the pf_queue_accountant_t must be fully cleared.
+ *
+ * @param p_accountant     InOut: Queue accountant
  */
-static int pf_alarm_reset_send_queues (pnet_t * net, pf_ar_t * p_ar)
+void pf_alarm_queue_mutex_create (pf_queue_accountant_t * p_accountant)
 {
-   memset (p_ar->alarm_send_q, 0, sizeof (p_ar->alarm_send_q));
+   if (p_accountant->mutex == NULL)
+   {
+      p_accountant->mutex = os_mutex_create();
+   }
+}
+
+/**
+ * @internal
+ * Destroy a mutex for a queue.
+ *
+ * @param p_accountant     InOut: Queue accountant
+ */
+void pf_alarm_queue_mutex_destroy (pf_queue_accountant_t * p_accountant)
+{
+   if (p_accountant->mutex != NULL)
+   {
+      os_mutex_destroy (p_accountant->mutex);
+      p_accountant->mutex = NULL;
+   }
+}
+
+/**
+ * @internal
+ * Reset the queue indexes.
+ *
+ * Locks the queue.
+ *
+ * @param p_accountant     InOut: Queue accountant
+ */
+static void pf_alarm_queue_lock_and_reset (pf_queue_accountant_t * p_accountant)
+{
+   assert (p_accountant->mutex != NULL);
+   os_mutex_lock (p_accountant->mutex);
+   p_accountant->count = 0;
+   p_accountant->read_index = 0;
+   p_accountant->write_index = 0;
+}
+
+/**
+ * @internal
+ * Unlock a queue.
+ *
+ * @param p_accountant     InOut: Queue accountant
+ */
+static void pf_alarm_queue_unlock (pf_queue_accountant_t * p_accountant)
+{
+   assert (p_accountant->mutex != NULL);
+   os_mutex_unlock (p_accountant->mutex);
+}
+
+/**
+ * @internal
+ * Get next write index for a queue.
+ *
+ * Locks the queue on success.
+ *
+ * @param p_accountant     InOut: Queue accountant
+ * @param p_write_index    Out:   Write index to use
+ * @return 0 if there is room in the queue (Queue will be locked)
+ *         -1 if the queue is full (Will not be locked)
+ */
+static int pf_alarm_queue_lock_and_get_writeindex (
+   pf_queue_accountant_t * p_accountant,
+   uint16_t * p_write_index)
+{
+   os_mutex_lock (p_accountant->mutex);
+   if (p_accountant->count >= PNET_MAX_ALARMS)
+   {
+      os_mutex_unlock (p_accountant->mutex);
+
+      return -1;
+   }
+
+   *p_write_index = p_accountant->write_index;
+
+   p_accountant->write_index++;
+   if (p_accountant->write_index >= PNET_MAX_ALARMS)
+   {
+      p_accountant->write_index = 0;
+   }
+   p_accountant->count++;
+
+   return 0;
+}
+
+/**
+ * @internal
+ * Get next read index for a queue.
+ *
+ * Locks the queue on success.
+ *
+ * @param p_accountant     InOut: Queue accountant
+ * @param p_read_index     Out:   Read index to use
+ * @return 0 if the queue has an item to read (Queue will be locked)
+ *         -1 if the queue is empty (Will not be locked)
+ */
+static int pf_alarm_queue_lock_and_get_readindex (
+   pf_queue_accountant_t * p_accountant,
+   uint16_t * p_read_index)
+{
+   // TODO Remove, find out why it segfaults sometimes.
```

Comment:
?

Reply:
Implemented logic to handle incoming frames before or after the queue is in its initialized state.
Done